### ISSUE TITLE
Drop Flask entirely for Django

### ISSUE THREAD

**Issue by biblicabeebli** [onnela-lab/beiwe-backend#255]
Issue for tracking the Django porting process

## This is an involved process, the reasons are many, this work is long-overdue
This issue tracks the work on the branch [`eli/django`](https://github.com/onnela-lab/beiwe-backend/tree/eli/django)

This development work is critical for the platform health of Beiwe in general.  By ditching the old flask and adopting django especially in combination with the new tests that are part of this work, the substantial maintenance burden will (should) be reduced.

- Flask is a limited framework, it does not have the bulk features of Django.
- Flask is less widespread than Django, it will open up development opportunities to more people.
- The presence of Flask and Django causes general confusion, esp in bug reports and to other collaborators attempting to contribute.
- We shifted to using the Django ORM layer some time ago.  The remaining work consists of updating website endpoints to use the django prrimitives, and dropping any explicit flask dependencies.  Some work has involved developing shims to copy flask features.
- There is an occasional problem with database connection issues, this stems from a django-only feature that opens and closes the database connection intelligently based on incoming requests.  This behavior causes occasionally broken deployments that can only be diagnosed by checking for the number of open database connections on the RDS server, the only known fix is to reboot the database, or wait until the connections timeout.
- security - flask is missing many security features that need to be implemented externally to the web framework, this will reduce sysadmin work/responsibility as we can finally automat a couple features.
- tests - building tests has long been problematic for the platform.  Running django-native allows a much easier test framework that the developers are already familiar with.  it also means we don't have to maintain our own weird compatibility layer in the tests.
- general dependency reduction.

## Current Status
- all _website_ endpoints have at least a first-pass transformation to using django, substantial testing is needed but many things load.  Quality is any of Alpha, Beta, or Release Candidate, depending on ease of testing, mostly Beta+.
- API endpoints are mostly Alpha, endpoints that are required for loading pages are RC.
- admin/researcher authentication is pretty good, Beta or RC..
- download api authentication needs testing and development, Alpha.
- tableau and its authentication is **not implemented**.
- mobile pages and device authentication are Alpha, first pass with zero testing.

## Tests
Tests have been moved to a top-level folder.  We are currently using the djangotestrunner, open to using additional libraries.
- we don't need 100% test code coverage of all endpoints, but all endpoints must have at least one test on them.
- all authentication code must have **100% code coverage before release.**
- all device endpoints must have **100% code coverage before release.**
- selenium tests will not be included in this work.
- javascript - no plans at this time, **looking for input**.
- the `factory_boy` dependency is currently in use only on the (preexisting) tableau API tests.  **Use of this library is currently considered an impediment to testing because it actively obscures the database aspects of the tests, making correctness difficult to determine.**  Further discussion on this topic is expected.

## How To Test Code Coverage Using The `coverage` Library
- you will need to pip install the `coverage` library into your project python environment.  There is a new requirements_testing.txt file.
- `coverage` is fairly simple to use, but it does generate files and you may forget to clean up after previous runs, here is a bash function that will allow you to easily run tests with coverage without having to keep details about it in your head.
- use the below function like this, documentation is inlined.  First parameter is the folder to run `coverage` over, all following parameters get passed into the test command.

`your_renamed_function authentication tests.test_endpoints.TestAuthentication`

```
function please_rename_me(){
    if (( $# != 2 ))
        then echo "need 2 params: folder to apply python coverage, and then tests to run (empty string for all tests)"
        return
    fi

    coverage erase  # erase existing coverage content
    rm -rf htmlcov  # delete the generated html report

    # coverage requires a folder target containing all python code you want the coverage report on to operate on.
    #   provide "." to run on all code, but that can be slow.
    # Executing a test with coverage takes longer than normal, recommend using the time directive to give you 
    #   sanity-saving time feedback so you can adapt your workflow.
    # The testing directive here is to by-default run with --keepdb for speed, and to run in extra verbose mode because otherwise print statements don't work and that can be aggrivating.  Change it if you dislike.
    
    # Capture the first parameter you passed to this function
    # `shift` `shifts all of this function's parameters down by one, deletes the contents of $1 that we grabbed above.
    #   The remaining parameters get passed directly to `manage.py test`, which is populated with some defaults that you should review.
    COVERAGE_FOLDER=$1
    shift

    time coverage run --source="$COVERAGE_FOLDER" manage.py test -v2 --keepdb $*

    # (this can take some time, so I've stuck a print statement)
    echo "\n" building html report
    coverage html

    # attempts to open the new coverage file in a way that works on mac and linux, your call whether to include.
    echo Opening...
    open -a '/Applications/Google Chrome.app/' htmlcov/index.html 2> /dev/null || xdg-open htmlcov/index.html
}
```

**Comment by biblicabeebli** [comment]
(Notification because I am mentioning this branch to other people now that it has good progress)

There has been a regression in functionality, I screwed up slightly on the exact implementation of the `url` jinja template extension, so a lot of `href="{{ url ... }} "` things are currently broken.  I'm hitting them and fixing them as I go through building tests.

**Comment by biblicabeebli** [comment]
Udpate for those paying attention:
The Django port is at roughly "Pretty much done." on the [Done Scale](http://20px.com/blog/2014/12/05/done-scale/).


- Every use of the Flask library has been removed from active code, Django can now run the server and every endpoint is present, many of them even work.
- References to Flask are still present, and no work has been done to match configuration present in `app.py`.
- There are now 50 tests of varying quality covering quite a few of the api and pages files.  I'm simply going down through the `urls` file and covering every endpoint.
- I have used `coverage` on a handful of the tests, I will try using it to fill in tests in more detail later on for critical items.  The goal right now is to cover every endpoint with the test client with a reasonable amount of confirmation that behavior is as intended, and that pages load and return the correct status codes.
- There are no unit tests, and no tests for code that is not directly served from the Flask app.  Such tests are not expected to be present before merging into `main`.
- I have scattered quite a few `FIXME`'s across the codebase to address many issues that should be cleaned up before merging to `main`, but only some of those are blockers.
- No testing work has been completed on any of the `mobile_api` code.

The testing abstractions are still in development, but the major patterns at play, excluding `mobile_api`, have been identified.  Documentation of the abstractions is pretty sparse as they could still change, but I will get around to documenting them.

**Comment by biblicabeebli** [comment]
### Udpate!
The Django port is at roughly "Done except for..." on the Done Scale.

- There are now 107 tests.
- _Many_ Django port issues have been resolved.
- Many bugs have been fixed.
- There are currently 57 `fixme`s across the codebase, about half of them attached to the existing tests noting additional cases to be implemented.
- No real update at this time with respect to use of Coverage to gauge the quality of tests.
- Have not gotten to tests for data download, tableau, or mobile api.  Those are the "hard ones", but fortunately all the expected changes have been made and "only the bugs" remain.
- The abstractions and code tools needed to make concise, legible tests for HTTP endpoints are largely complete.
- The abstractions are mostly documented!
- The tools for generating appropriate database objects for testing purposes are still growing, they are located in `tests/helpers.py`.
- Requirements have been updated, testing requirements have been updated.
- The branch has 139 commits and is up-to-date with code changes on `main`.

Goal is to get this on our staging server before Thanksgiving day in the US (November 25).

**Comment by biblicabeebli** [comment]
Moderate progress this week, but the tedious part of bulk test building and "does page render without crashing" are finished. 154 commits total.

I've built tests for every endpoint that is part of the normal researcher login session, moving on to the endpoints that require alternate credentials, or come from the mobile apps.

Done scale progress: unchanged, still "Done except for..."

**Comment by biblicabeebli** [comment]
Data access api now has 100% test coverage (technically 98%) due to some debugging behavior.

**Comment by biblicabeebli** [comment]
Mobile api now has test coverage.
- The ios endpoints are weird, they use http basic auth for bad reasons, and I have not worked out how to test that detail yet, but other than weird authentication behavior they are literally the same code.
- Device upload does not yet have a success testcase, that is particularly hairy code to "just build" valid data for, so I will finish that test when I get to real device testing, which will be... soon.  (This week? :D )

@alvinsiu I have ported your tests for tableau stuff, thanks for putting those together.  They may be unrecognizable.

The Massive 2,415 line test_endpoints.py file will be factored out into some yet-to-be-determined organization, I have avoided doing this because imho there are a **lot** of endpoints that should also be moved around, and I have implemented two meta tests to help me track these organizational details.

### You can now just run all tests and the experience is reasonable.
- TESTS PASS AND THAT ACTUALLY MEANS SOMETHING!
- Test coverage in `api` and `pages` are 69% and 76%.  This will increase some as the final tests are put together and some fixmes are addressed in the existing tests.  100% coverage is not a goal at this time.  `libs` only has 28% coverage, but it contains developer utils not accessed in the live codebase, and components that are unreachable from the frontend.
- The two aforementioned meta tests are currently skipped (they do not pass right now).
- All tests can be run in parallel, just pass in `--parallel number_of_cpu_cores`.  The test architecture is expected to maintain this to allow for a fast iteration process for developers in the future, and is recommended.
- Strongly recommend running with `--keepdb` parameter, there are 62 migrations.
- There is extra terminal output on verbosity level 2 or greater.  Output from tests are bracketed with `==` and new lines. Calls to the messages library (e.g. `messages.error`, etc.) will additionally print their output to the terminal to assist with test development.
- Calls that would hit s3 in the `libs.s3` functions will automatically fail with a(n insane) Boto3 error.  This is accomplished by patching the Bucket name to be a None.  This is the case for all tests and cannot be disabled.  Tests that require S3 functionality must use the mock library's patch functionality.  It is not a goal right now to substantially test `libs.s3`.
- requirements have been updated, and a requirements_testing.txt file is present that includes the `mock` and `coverage` libraries, as well as `tblib`, which is necessary to serialize traceback objects when they occur in parallel mode.
- All work has been to be compatible with Pythons 3.6, 3.7, and 3.8.  **Note that the current `main` branch is not compatible with 3.8**

@zagorsky given the sheer breadth of the work and the presence of the new tests I will not be asking for intensive review, but this work is substantially progressed and I intend to have it on at least the staging server this week.  Please take a look at the branch, you may want a new python env, and a new python version.  As soon as this is done (or at least on staging for testing) I will be starting on the Python 3.8 platform upgrade work.  I have not looked at the work on the messaging branch at all.  If you want to take a whack at that merge/rebase you should do it as a merge into [a copy of] the `eli/django` branch; the rebase is painful, there are moved and renamed files, import conflicts, copy-pastes of code, etc.

**Comment by biblicabeebli** [comment]
Update:
Done scale: Done v3
We are live on our staging server.
iOS support is broken.
Everything else seems to work.

**Comment by biblicabeebli** [comment]
and now iOS should work.  Unless I've misunderstood something about http basic auth

**Comment by biblicabeebli** [comment]
We are dogfooding this now on all the Onnela Lab servers.

**Comment by biblicabeebli** [comment]
This work is finished, only remaining items are a merge and a possible requirements update for forest.
